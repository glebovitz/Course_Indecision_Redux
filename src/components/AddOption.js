import React from 'react';

export default class AddOption extends React.Component {
    state = {
        inputValue: '',
        optionCount: 0 //used to store previous optionCount prop
    };

    // used changes in optionCount to determine that we successfully added
    // a new option. Previous optionCount is stored in local state.
    static getDerivedStateFromProps = (nextProps, prevState) => {
        if (nextProps.optionCount !== prevState.optionCount) {
            return { 
                inputValue: '',
                optionCount: nextProps.optionCount,
            };
        } else {
            return null;
        }
    };
    handleAddOption = (e) => {
        e.preventDefault();
        this.props.addOption(this.state.inputValue);
    };
    handleOnChange = (e) => {
        const inputValue = e.target.value;
        this.setState(() => ({
            inputValue
        }));
    };
  render() {
      return (
        <div>
            {!!this.props.error && <p className='add-option-error'>{this.props.error}</p>}
            <form className='add-option' onSubmit={this.handleAddOption}>
                <input 
                    className='add-option__input'
                    type='text'
                    value={this.state.inputValue}
                    onChange={this.handleOnChange}
                />
                <button className='button'>Add Option</button>
            </form>
        </div>
      );
  }
}
