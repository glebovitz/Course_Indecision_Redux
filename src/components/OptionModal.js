import React from 'react';
import Modal from 'react-modal';

export default class OptionModal extends React.Component {
  state = {
    lastSelectedOption: ''
  };

  // Sequence: 
  //    1. props.selectedOption is cleared. 
  //    2. Modal is content updated.
  //    3. Modal then closes.
  // This causes the dialog button to jump upward.
  // Instead store selected option in a local state variabe and only change
  // when we have a new non empty selected option.
  static getDerivedStateFromProps = (nextProps, prevState) => {
    if (nextProps.selectedOption) {
        return { 
            lastSelectedOption: nextProps.selectedOption
        };
    } else {
        return null;
    }
  };

  render() {
    return (
      <Modal
        isOpen={!!this.props.selectedOption}
        onRequestClose={() => this.props.clearSelectedOption()}
        contentLabel="Selected Option"
        closeTimeoutMS={200}
        className='modal'
        ariaHideApp={false}
      >
        <h3 className='modal__title' >Select Option</h3>
        {this.state.lastSelectedOption && 
          <p className='model__body' >{this.state.lastSelectedOption}</p>}
        <button  className='button' onClick={() => this.props.clearSelectedOption()}>
          Okay
        </button>
      </Modal>
    );
  };
};
