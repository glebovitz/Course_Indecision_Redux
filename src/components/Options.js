import React from 'react';
import { connect } from 'react-redux';
import Option from './Option';
import { deleteOptions, deleteOption } from '../actions/options';

export const Options = (props) => (
    <div>
        <div className="widget-header">
            <h3 className="widget-header__title">Your Options</h3>
            <button 
                className="button button--link"
                onClick={props.deleteOptions}
            >
                Remove All
            </button>
        </div>
        {props.options.length === 0 && <p className="widget__message">Please Add and Option to get started</p>}
        {
            props.options.map((option, index) => (
                <Option 
                    key={option} 
                    optionText={option}
                    count={index + 1}
                    deleteOption={props.deleteOption}
                />
            ))
        }
    </div>
);

const mapDispatchToProps = (dispatch) => ({
    deleteOptions: () => dispatch(deleteOptions()),
    deleteOption: (option) => dispatch(deleteOption(option))
});

const mapStateToProps = (state) => {
    return {
        options: state.options.options
    };
};

export default connect(mapStateToProps, mapDispatchToProps)(Options);
