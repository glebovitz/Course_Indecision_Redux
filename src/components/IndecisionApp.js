import React from 'react';
import { connect } from 'react-redux';
import AddOption from './AddOption';
import Header from './Header';
import Action from './Action';
import Options from './Options';
import OptionModal from './OptionModal';
import { 
    setOptions, 
    selectOption,
    addOption,
    clearSelectedOption 
} from '../actions/options';

export class IndecisionApp extends React.Component {
    componentDidMount() {
        try {
            const json = localStorage.getItem('options');
            const options = JSON.parse(json);
            if (options) {
                this.props.setOptions(options);
            };
        } catch (e) {
            // NOTHING
        };
    };
    componentDidUpdate(prevProps, prevState) {
        if (prevProps.options.length !== this.props.options.length) {
            const json = JSON.stringify(this.props.options);
            localStorage.setItem('options', json);
        }
    };
    pickOption = () => {
        const randomNum = Math.floor(Math.random() * this.props.options.length);
        const option = this.props.options[randomNum];
        this.props.selectOption(option);
    };
    render () {
        const title = "Indecision";
        const subtitle = "Put you life in the hands of a computer.";
        return (
            <div>
                <Header 
                    title={title} 
                    subtitle={subtitle}
                />
                <div className='container'>
                    <Action 
                        hasOptions={this.props.options.length > 0}
                        pickOption={this.pickOption}
                    />
                    <div className='widget'>
                    <Options />
                    <AddOption
                        addOption={this.props.addOption}
                        optionCount={this.props.options.length}
                        error={this.props.error}
                    />
                    </div>
                </div>
                <OptionModal
                    selectedOption={this.props.selectedOption}
                    clearSelectedOption={this.props.clearSelectedOption}
                />
            </div>
        );
    };
};

const mapStateToProps = (state) => {
    return {
        options: state.options.options,
        selectedOption: state.options.selectedOption,
        error: state.options.error
    };
};

const mapDispatchToProps = (dispatch) => ({
    setOptions: (options) => dispatch(setOptions(options)),
    selectOption: (option) => dispatch(selectOption(option)),
    addOption: (option) => dispatch(addOption(option)),
    clearSelectedOption: () => dispatch(clearSelectedOption())
});

export default connect(mapStateToProps, mapDispatchToProps)(IndecisionApp);
