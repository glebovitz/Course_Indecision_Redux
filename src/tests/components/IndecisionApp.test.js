import React from 'react';
import { shallow, mount } from 'enzyme';
import { IndecisionApp } from '../../components/IndecisionApp';
import testOptions from '../fixtures/options';

test('should render Component with values correctly', () => {
  const wrapper = shallow (
    <IndecisionApp
      options={testOptions}
      selectedOption={'testabcdef'}
      error={'test123456'}
    />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Component with defaults correctly', () => {
  const wrapper = shallow (
    <IndecisionApp
      options={[]}
      selectedOption={undefined}
      error={undefined}
    />);
  expect(wrapper).toMatchSnapshot();
});

test('should call selectOption from pickOption correctly', () => {
  const selectOption = jest.fn();
  const wrapper = shallow (
    <IndecisionApp
      options={testOptions}
      selectedOptions={undefined}
      selectOption={selectOption}
      error={undefined}
    />
  );
  wrapper.find("Action").props().pickOption();
  expect(selectOption).toHaveBeenCalled();
});

test('should call selectOption from pickOption correctly', () => {
  const addOption = jest.fn();
  const value = "testabcdef";
  const wrapper = shallow (
    <IndecisionApp
      options={testOptions}
      selectedOptions={undefined}
      addOption={addOption}
      error={undefined}
    />
  );
  wrapper.find("AddOption").props().addOption();
  expect(addOption).toHaveBeenCalled();
});

test('should call clearSelectedOption to be called correctly from OptionModal', () => {
  const clearSelectedOption = jest.fn();
  const value = "testabcdef";
  const wrapper = shallow (
    <IndecisionApp
      options={testOptions}
      selectedOptions={undefined}
      clearSelectedOption={clearSelectedOption}
      error={undefined}
    />
  );
  wrapper.find("OptionModal").props().clearSelectedOption();
  expect(clearSelectedOption).toHaveBeenCalled();
});

