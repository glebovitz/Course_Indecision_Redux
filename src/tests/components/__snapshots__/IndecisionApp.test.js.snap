// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Component with defaults correctly 1`] = `
<div>
  <Header
    subtitle="Put you life in the hands of a computer."
    title="Indecision"
  />
  <div
    className="container"
  >
    <Action
      hasOptions={false}
      pickOption={[Function]}
    />
    <div
      className="widget"
    >
      <Connect(Options) />
      <AddOption
        optionCount={0}
      />
    </div>
  </div>
  <OptionModal />
</div>
`;

exports[`should render Component with values correctly 1`] = `
<div>
  <Header
    subtitle="Put you life in the hands of a computer."
    title="Indecision"
  />
  <div
    className="container"
  >
    <Action
      hasOptions={true}
      pickOption={[Function]}
    />
    <div
      className="widget"
    >
      <Connect(Options) />
      <AddOption
        error="test123456"
        optionCount={3}
      />
    </div>
  </div>
  <OptionModal
    selectedOption="testabcdef"
  />
</div>
`;
